util/gpool/spmc/worker.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spmc
+
+import (
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/resourcemanager/pooltask"
+	"go.uber.org/zap"
+)
+
+// goWorker is the actual executor who runs the tasks,
+// it starts a goroutine that accepts tasks and
+// performs function calls.
+type goWorker[T any, U any, C any, CT any, TF pooltask.Context[CT]] struct {
+	// pool who owns this worker.
+	pool *Pool[T, U, C, CT, TF]
+
+	// taskBoxCh is a job should be done.
+	taskBoxCh chan *pooltask.TaskBox[T, U, C, CT, TF]
+
+	exit chan struct{}
+
+	// recycleTime will be updated when putting a worker back into queue.
+	recycleTime time.Time
+}
+
+// run starts a goroutine to repeat the process
+// that performs the function calls.
+func (w *goWorker[T, U, C, CT, TF]) run() {
+	w.pool.addRunning(1)
+	go func() {
+		defer func() {
+			w.pool.addRunning(-1)
+			w.pool.workerCache.Put(w)
```

Comment:
Could we put this to https://github.com/pingcap/tidb/pull/39872/files#diff-526c36d8ad42670e920fd84ae336d6b60196c61639bfa1c3a572adad64d1836cR328? Then I think it's easier to understand and it matches